modules/service/src/main/scala/lucuma/odb/service/ProgramService.scala

```diff
@@ -130,29 +163,54 @@ object ProgramService {
         }
       }
 
-      def updatePrograms(props: ProgramPropertiesInput, where: AppliedFragment): F[Unit] =
-        Statements.updatePrograms(props, where).traverse { af =>
-          s.prepare(af.fragment.command).use(_.execute(af.argument))
-        } .void
+      def updatePrograms(SET: ProgramPropertiesInput.Edit, where: AppliedFragment): F[UpdateProgramResponse] =
+        s.transaction.use { xa =>
+
+          // Update the program itself
+          val a: F[UpdateProgramResponse] =
+            Statements.updatePrograms(SET, where).fold(UpdateProgramResponse.Success(Nil).pure[F]) { af =>
+              s.prepare(af.fragment.query(program_id)).use { ps =>
+                ps.stream(af.argument, 1024)
+                  .compile
+                  .toList
+                  .map(UpdateProgramResponse.Success(_))
+              }
+            }
+
+          // Update the proposal
+          val b: F[UpdateProgramResponse] =
+            SET.proposal.fold(UpdateProgramResponse.Success(Nil).pure[F]) {
+              proposalService.updateProposals(_, where, xa).map {
```

Comment:
Will the `where` results differ here potentially from the `updatePrograms` case?

Reply:
Yes, there may not be any updates to program properties at al, in which case nothing is done by `a` above, which is also the case here. If both the program fields and the proposal fields are updated then yes, they should yield the same set of program ids.

Reply:
i guess what i mean is, when the `where` fragment does its select it may end up selecting _different_ programs than those selected in the `a` case.  For example, if you select ones titled "Biff" and update them to "Buff" it will no longer find any "Biff"s.  Does that make sense?  Or am I reading it wrong?

Reply:
Actually the way to do this is to select them into a temporary table and join to it. That way we're not sending the ids all over the place.

Reply:
Ok I changed it to write a temp table containing the program ids selected by `where`, plus an indication of whether there is an existing proposal or not. Then I join with it to do all the follow-on updates. Still adding tests so I'll keep this unresolved for now.